feat!: introduce confidence scores for check facts (#620)
oracle/macaron: docs/source/pages/developers_guide/index.rst

For API reference, see the :doc:`API Reference </pages/developers_guide/apidoc/index>` page.

-------------------
Writing a New Check
-------------------

> 

Reviewer: I think all mid-paragraph line breaks should be removed from this file.

Author: Hm, I have added them to help with git diffs.

As a contributor to Macaron, it is very likely to need to write a new check or modify an existing one at some point. In this
section, we will understand how Macaron checks work and what we need to do to develop one.

+++++++++++++++++
High-level Design
+++++++++++++++++

Before jumping into coding, it is useful to understand how Macaron as a framework works. Macaron is an extensible
framework designed to make writing new supply chain security analyses easy. It provides an interface
that you can leverage to access existing models and abstractions instead of implementing everything from scratch. For

> instance, many security checks require to traverse through the code in GitHub Actions configurations. Normally,

Reviewer: ```suggestion
instance, many security checks require traversing through the code in GitHub Actions configurations. Normally,
```

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

you would need to find the right repository and commit, clone it, find the workflows, and parse them. With Macaron,
you don't need to do any of that and can simply write your security check by using the parsed shell scripts that are
triggered in the CI.

Another important aspect of our design is that all the check results are automatically mapped and stored in a local database.

> By performing this mapping, we make it possible to enforce flexible policies on the results of the checks. While storing

Reviewer 2: Nit: The meaning of the word `flexible` in `flexible policies` here feels vague to me. I would prefer something along the line of:

```suggestion
By performing this mapping, we make it possible for users to enforce their use case-specific policies on the results of the checks. While storing
```

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

> the check results to the database happens automatically by Macaron's backend, the developer needs to add a brief specification

Reviewer: ```suggestion
the check results to the database happens automatically in Macaron's backend, the developer needs to add a brief specification
```

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

to make that possible as we will see later.

> 

Reviewer 2: I think before elaborating on how to implement a check, we may want to clarify at a high level what a check does. Some simple high-level questions that may come to mind for a new contributor could be:
* What are the inputs of a check?
* What are the outputs of a check?
* Is it safe to assume that each check should leave no side-effects, i.e. does it only read the inputs and return the outputs, or does it modify a shared state with other checks?

Some other questions that are not as high-level but still carry a high-level understanding of the checks include:
* Do the checks run in a particular order? What is the meaning of `depends_on` and how does the check dependency relationship work?
* When are these checks executed in Macaron? What does Macaron do before and after the running of the checks? Here, it could even be helpful to emphasize that these checks run as part of `macaron analyze`.

At the same time, it could also be helpful to not just clarify what should be implemented as a check, but also what **should not** be implemented as a check, and one should leverage the policy engine instead. I feel like the `RepoCheck` example that you gave is not particularly representative of what should be implemented as a check since Macaron has already produced this information in the database and one should just write a policy instead of implementing a new check. If you think this is the case, we can either provide a different example, or simply say that "This is just an example for how a check should be implemented, and not an example of what a check should be for. Please open an issue to discuss with us if you want to implement a new check and we can confirm whether your use case is appropriate."

Author: Thanks for the suggestions. I will incorporate them in the documentation. A the same time, I want to keep the description short but useful.

This example check is not meant to be representative and just shows how to create a new check.  More information on what a check should not be is better to be added in a FAQ section in a different PR. Drawing a line between a check and policy and discussing what should be part of a check vs policy is also a topic for a FAQ section.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

+++++++++++++++++++
The Check Interface
+++++++++++++++++++

Each check needs to be implemented as a Python class in a Python module under ``src/macaron/slsa_analyzer/checks``.

> A check class should subclass the ``BaseCheck`` class in :ref:`base_check module <pages/developers_guide/apidoc/macaron\.slsa_analyzer\.checks:macaron.slsa\\_analyzer.checks.base\\_check module>`.

Reviewer 2: You can use [these Sphinx Python roles](https://www.sphinx-doc.org/en/master/usage/domains/python.html#python-roles) to cross-reference Python objects, which is nicer than using the `:ref:` role.
```suggestion
A check class should subclass the :class:`BaseCheck <macaron.slsa_analyzer.checks.base_check.BaseCheck>` class.
```

Author: I was looking for a way to reference directly. That's great thanks.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

> You need to set the name, description, and other details of your new check in the ``__init__`` method. After implementing

Reviewer 2: Here, by `name` I assume you are referring to the check id. Do you think we should mention the naming rule for checks?
The docstring of the `Registry._validate_check_id_format()` method says:
```
        A valid check id must have the following properties:
            - The general format: ``mcn_<name>_<digits>``
            - In ``name``, only lowercase alphabetical letters are allowed. If ``name`` contains multiple \
            words, they must be separated by underscores.
```
I ran into this issue once myself, and a new contributor is likely to run into the same problem of the registry not accepting their check if we do not mention it. On the other hand, if the naming has a convention, I think we should mention it.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

the initializer, you need to implement the ``run_check`` abstract method. This method provides the context object

> :ref:`AnalyzeContext <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.analyze\\_context module>`, which contains various

Reviewer 2: ```suggestion
:class:`AnalyzeContext <macaron.slsa_analyzer.analyze_context.AnalyzeContext>`, which contains various
```

intermediate representations and models. The ``dynamic_data`` property would be particularly useful as it contains

> data about the CI service, artifact registry, and build tool used for building the software component.

Reviewer 2: Some example code snippets should be helpful for this part.

Reviewer 2: To clarify, what I mean here is something like

```
.. code-block:: python
   import ...

   class MyCheck(BaseCheck):
       ...
```

You have some other example snippets after this part in the same section, so it makes sense also to do that here.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

> ``component`` is another useful attribute in the :ref:`AnalyzeContext <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.analyze\\_context module>` object

Reviewer: ```suggestion
The ``component`` is another useful attribute in the :ref:`AnalyzeContext <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.analyze\\_context module>` object
```

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

> ``component`` is another useful attribute in the :ref:`AnalyzeContext <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.analyze\\_context module>` object

Reviewer: I understand the use of `attribute` and `property` is intentional. However, given that the two differ slightly, using "another" as per "another useful attribute" does not follow. I would instead suggest:
```suggestion
The :ref:`AnalyzeContext <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.analyze\\_context module>` object contains other useful information, such as ``component``. This acts as a representation of a software component, including it's corresponding ``repository`` and ``dependencies``.
```
Another alternative would be to use the same word for `property` and `attribute` depending on which is more general.
Or perhaps I am simply being too pedantic in this case.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

> ``component`` is another useful attribute in the :ref:`AnalyzeContext <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.analyze\\_context module>` object

Reviewer 2: ```suggestion
``component`` is another useful attribute in the :class:`AnalyzeContext <macaron.slsa_analyzer.analyze_context.AnalyzeContext>` object
```

that you should know about. This attribute contains the information about a software component, such
as it's corresponding ``repository`` and ``dependencies``. Note that ``component`` will also be stored into the database and its attributes
such as ``repository`` are established as database relationships. You can see the existing tables and their
relationships in our :ref:`data model <pages/developers_guide/apidoc/macaron.database:macaron.database.table\\_definitions module>`.

Once you implement the logic of your check in the ``run_check`` method, you need to add a class to help
Macaron handle your check's output:

>    * Add a class that subclasses ``CheckFacts`` to map your outputs to a table in the database. The class name should follow the ``<MyCheck>Facts`` pattern.

Reviewer 2: ```suggestion
   * Add a class that subclasses :class:`CheckFacts <macaron.database.table_definitions>` to map your outputs to a table in the database. The class name should follow the ``<MyCheck>Facts`` pattern.
```

   * Specify the table name in the ``__tablename__ = "_my_check"`` class variable. Note that the table name should start with ``_`` and it should not have been used by other checks.
   * Add the ``id`` column as the primary key where the foreign key is ``_check_facts.id``.
   * Add columns for the check outputs that you would like to store into the database. If a column needs to appear as a justification in the HTML/JSON report, pass ``info={"justification": JustificationType.<TEXT or HREF>}`` to the column mapper.
   * Add ``__mapper_args__`` class variable and set ``"polymorphic_identity"`` key to the table name.

Next, you need to create a ``result_tables`` list and append check facts as part of the ``run_check`` implementation.

> You should also specify a :ref:`Confidence <pages/developers_guide/apidoc/macaron\.slsa_analyzer\.checks:macaron.slsa\\_analyzer.checks.check\\_result module>`
> score choosing one of  the ``Confidence`` enum values, e.g., ``Confidence.HIGH`` and pass it via keyword

Reviewer 2: ```suggestion
You should also specify a confidence score by choosing one of  the :class:`Confidence <macaron.slsa_analyzer.checks.check_result.Confidence>` enum values, e.g., :attr:`Confidence.HIGH <macaron.slsa_analyzer.checks.check_result.Confidence.HIGH>` and pass it via keyword
```

argument ``confidence``. You should choose a suitable confidence score based on the accuracy
of your check analysis.

.. code-block:: python

   result_tables.append(MyCheckFacts(col_foo=foo, col_bar=bar, confidence=Confidence.HIGH))

This list as well as the check result status should be stored in a :ref:`CheckResultData <pages/developers_guide/apidoc/macaron\.slsa_analyzer\.checks:macaron.slsa\\_analyzer.checks.check\\_result module>`
object and returned by ``run_check``.

> Finally, you need to register your check by adding it to the :ref:`registry module <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.registry module>`:

Reviewer 2: ```suggestion
Finally, you need to register your check by adding it to the :mod:`registry<macaron.slsa_analyzer.registry>` module:
```

> Finally, you need to register your check by adding it to the :ref:`registry module <pages/developers_guide/apidoc/macaron\.slsa_analyzer:macaron.slsa\\_analyzer.registry module>`:

Reviewer 2: For clarity, I think we should mention that this should be done at the end of the same file.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

.. code-block:: python

   registry.register(MyCheck())

And of course, make sure to add tests for you check by adding a module under ``tests/slsa_analyzer/checks/``.

+++++++
Example
+++++++

In this example, we show how to add a check determine if a software component has a source-code repository.
Feel free to explore other existing checks under ``src/macaron/slsa_analyzer/checks`` for more examples.

1. First create a module called ``repo_check.py`` under ``src/macaron/slsa_analyzer/checks``.

> 2. Add a class and specify the columns that you want to store for the check outputs to the database.

Reviewer: ```suggestion
2. Add a class and specify the columns that you want to store for the check outputs in the database.
```

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

           ]  # Choose a SLSA requirement that roughly matches this check from the ReqName enum class.
           super().__init__(check_id=check_id, description=description, depends_on=depends_on, eval_reqs=eval_reqs)

       def run_check(self, ctx: AnalyzeContext) -> CheckResultData:
           """Implement the check in this method.

           Parameters
           ----------
           ctx : AnalyzeContext
                 The object containing processed data for the target software component.

           Returns
           -------
           CheckResultData
                 The result of the check.
           """
           if not ctx.component.repository:
               logger.info("Unable to find a Git repository for %s", ctx.component.purl)
               # We do not store any results in the database if a check fails. So, just leave result_tables empty.
               return CheckResultData(result_tables=[], result_type=CheckResultType.FAILED)

           return CheckResultData(
               result_tables=[RepoCheckFacts(git_repo=ctx.component.repository.remote_path, confidence=Confidence.HIGH)],
               result_type=CheckResultType.PASSED,
           )

> 4. Register your check.

Reviewer 2: Repeating one of my previous comments: For clarity, I think we should mention that this should be done at the end of the same file.

Author: See commit [d25bf57](https://github.com/oracle/macaron/pull/620/commits/d25bf57f3e6f57d29cf4c311207dbcfe1302727e).

For API reference, see the :doc:`API Reference </pages/developers_guide/apidoc/index>` page.

-------------------
Writing a New Check
-------------------

> As a contributor to Macaron, it is very likely to need to write a new check or modify an existing one at some point. In this

Reviewer: ```suggestion
Contributors to Macaron are very likely to need to write a new check or modify an existing one at some point. In this
```

Author: See commit [08eb1ce](https://github.com/oracle/macaron/pull/620/commits/08eb1ce77c142753c5fae77b68d5b2b765c50e3d).

> section, we will understand how Macaron checks work and what we need to do to develop one.

Reviewer: ```suggestion
section, we will explain how Macaron checks work and what needs to be done by contributors working with them.
```

Author: See commit [08eb1ce](https://github.com/oracle/macaron/pull/620/commits/08eb1ce77c142753c5fae77b68d5b2b765c50e3d).

that you can leverage to access existing models and abstractions instead of implementing everything from scratch. For
instance, many security checks require traversing through the code in GitHub Actions configurations. Normally,
you would need to find the right repository and commit, clone it, find the workflows, and parse them. With Macaron,
you don't need to do any of that and can simply write your security check by using the parsed shell scripts that are
triggered in the CI.

Another important aspect of our design is that all the check results are automatically mapped and stored in a local database.
By performing this mapping, we make it possible to enforce use case-specific policies on the results of the checks. While storing
the check results in the database happens automatically in Macaron's backend, the developer needs to add a brief specification
to make that possible as we will see later.

Once you get familiar with writing a basic check, you can explore the check dependency feature in Macaron. The checks
in our framework can be customized to only run if another check has run and returned a specific

> :class:`result type <macaron.slsa_analyzer.checks.check_result.CheckResultType>`. This feature can be used when some checks

Reviewer: ```suggestion
:class:`result type <macaron.slsa_analyzer.checks.check_result.CheckResultType>`. This feature can be used when checks
```

Author: See commit [08eb1ce](https://github.com/oracle/macaron/pull/620/commits/08eb1ce77c142753c5fae77b68d5b2b765c50e3d).

> can be ordered and have a parent-child relationship, i.e., one check implements a weaker or stronger version of a

Reviewer: ```suggestion
have an ordering and a parent-child relationship, i.e., one check implements a weaker or stronger version of a
```

Author: See commit [08eb1ce](https://github.com/oracle/macaron/pull/620/commits/08eb1ce77c142753c5fae77b68d5b2b765c50e3d).

       git_repo: Mapped[str] = mapped_column(String, nullable=True, info={"justification": JustificationType.HREF})

       __mapper_args__ = {
           "polymorphic_identity": "_repo_check",
       }

'''''''''''''''''''
Add the check class
'''''''''''''''''''

Add a class for your check that subclasses :class:`BaseCheck <macaron.slsa_analyzer.checks.base_check.BaseCheck>`,
provide the check details in the initializer method, and implement the logic of the check in
:func:`run_check <macaron.slsa_analyzer.checks.base_check.BaseCheck.run_check>`.

A ``check_id`` should meet the following requirements:

    - The general format: ``mcn_<name>_<digits>``

>     - In ``name``, only lowercase alphabetical letters are allowed. If ``name`` contains multiple \
>         words, they must be separated by underscores.

Reviewer 2: This currently renders like this on my side:
![image](https://github.com/oracle/macaron/assets/49882031/ec8953bc-5840-4de1-87f6-ad2529b70478)

I suspect this is due to the additional indentation on the second line relative to the first, so Sphinx interpreted this as a [definition list item](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#definition-lists). This is instead a [bullet list item](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#bullet-lists) item, so the second line should be indented at the same level as the first line. The backslash at the end of the first line is also not necessary.

```suggestion
    - In ``name``, only lowercase alphabetical letters are allowed. If ``name`` contains multiple
      words, they must be separated by underscores.
```

Author: > This currently renders like this on my side

See comit [4661c00](https://github.com/oracle/macaron/pull/620/commits/4661c001a72ce289b550fe11591e031f331f943c).

A ``check_id`` should meet the following requirements:

    - The general format: ``mcn_<name>_<digits>``.

>     - Lowercase alphabetical letters are allowed in ``name`` only. If ``name`` contains multiple words, they must be separated by underscores.

Reviewer 2: Sorry for nitpicking but this sounds more like "the only part of the check id where lowercase alphabetical letters are allowed is `name`", which is different from "Only lowercase alphabetical letters are allowed in ``name``."

Author: When `only` is used at the end of a sentence, it's a modifier for the whole sentence. Here I don't want to focus on `name` or `lowercase`.

Reviewer 3: I agree with Nathan's interpretation of the literal meaning of the sentence. However, "Only lowercase alphabetical letters are allowed in `name`" would also be incorrect because it is immediately contradicted by the following statement that implies that underscore is also allowed.

Presuming that the intention is to convey that name has the form `[a-z]+(_[a-z]+)+`, a precise description would be "`name` must consist solely of words (sequences of lowercase alphabetical letters) separated by underscores" (you could even include the regex if you want to be absolutely precise).

(To be extra nitpicky, "in name only" is a such a common phrase with a specific meaning that I'd avoid invoking it by coincidence, I know that `name` is a specially-quoted identifier that shouldn't be read literally, but it still throws me a little)

Author: See commit [08eb1ce](https://github.com/oracle/macaron/pull/620/commits/08eb1ce77c142753c5fae77b68d5b2b765c50e3d).
